compiler/src/parsing/parser.mly

```diff
@@ -0,0 +1,707 @@
+include Parser_header
+
+(* https://github.com/ocaml/dune/issues/2450 *)
+module Grain_parsing = struct end
+%}
+
+
+%token <string> NUMBER_INT
+%token <string> NUMBER_FLOAT
+%token <string> INT32
+%token <string> INT64
+%token <string> FLOAT32
+%token <string> FLOAT64
+%token <string> WASMI32
+%token <string> WASMI64
+%token <string> WASMF32
+%token <string> WASMF64
+%token <string> ID
+%token <string> TYPEID
+%token <string> STRING
+%token <string> CHAR
+%token LBRACK LBRACKRCARET RBRACK LPAREN RPAREN LBRACE RBRACE LCARET RCARET
+%token LCARETLCARET RCARETRCARET RCARETRCARETRCARET
+%token CARET
+%token COMMA SEMI AS
+%token THICKARROW ARROW
+%token IS ISNT EQEQ LESSEQ GREATEREQ
+%token NOTEQ EQUAL GETS
+%token UNDERSCORE
+%token COLON DOT ELLIPSIS
+
+%token ASSERT FAIL EXCEPTION THROW
+
+%token PLUS PLUSPLUS DASH STAR SLASH PERCENT
+%token PLUSEQ DASHEQ STAREQ SLASHEQ PERCENTEQ
+%token TRUE FALSE VOID
+
+%token LET MUT REC IF WHEN ELSE MATCH WHILE FOR CONTINUE BREAK
+%token AMPAMP PIPEPIPE NOT AT
+%token AMP PIPE
+
+%token ENUM RECORD TYPE IMPORT EXPORT FOREIGN WASM PRIMITIVE
+%token EXCEPT FROM
+%token EOL EOF
+
+// reserved tokens
+%token TRY CATCH COLONCOLON
+
+// Not a real token, this is injected by the lexer
+%token FUN
+
+/* Operator precedence may be found in /docs/contributor/operator_precedence.md */
+
+%nonassoc _below_infix
+
+%left PIPEPIPE
+%left AMPAMP
+%left PIPE
+%left CARET
+%left AMP
+%left EQEQ NOTEQ IS ISNT
+%left LCARET LESSEQ RCARET GREATEREQ
+%left LCARETLCARET
+%left PLUS DASH PLUSPLUS
+%left STAR SLASH PERCENT
+
+%right SEMI EOL COMMA DOT COLON
+
+%nonassoc _if
+%nonassoc ELSE
+
+
+%start <Parsetree.parsed_program> program
+
+// This causes a syntax error to bubble up to a higher construct.
+// This helps provide more useful error messages, e.g. with the
+// expression `foo(bar baz` you'll get something like "Expected a
+// closing paren for the function call" rather than "A plus sign
+// could work here"
+%on_error_reduce
+  eols
+  record_get
+  array_get
+  non_assign_expr
+  annotated_expr
+  simple_expr
+  binop_expr
+  expr
+  typ
+  data_typ
+  id
+  id_expr
+  lbrace
+  lbrack
+  lbrackrcaret
+  lparen
+  lcaret
+  comma
+  eos
+  arrow
+  thickarrow
+  equal
+  const
+  pattern
+  type_id
+  value_binds
+
+%%
+
+// List helpers. These will parse a left-associated list.
+// Solves the shift/reduce conflicts you get from Menhir's
+// built-in lists.
+
+lseparated_nonempty_list_inner(sep, X):
+  | lseparated_nonempty_list_inner(sep, X) sep X { $3::$1 }
+  | X { [$1] }
+
+%inline lseparated_nonempty_list(sep, X):
+  | lseparated_nonempty_list_inner(sep, X) { List.rev $1 }
+
+%inline lseparated_list(sep, X):
+  | { [] }
+  | lseparated_nonempty_list(sep, X) { $1 }
+
+// The Grain grammar.
+
+%inline eol:
+  | EOL {}
+
+eols:
+  | eol+ {}
+
+%inline opt_eols:
+  | ioption(eols) {}
+
+eos:
+  | eols {}
+  | SEMI opt_eols %prec SEMI {}
```

Comment:
Any tips for understanding what these `%prec`s do? I looked at the manual entry, but I'm struggling with the intuition.

Reply:
`%prec expr` makes the reduction happen with the precedence of `expr`, whether that's a real token or a fake one that exists just for precedence. This one should be unnecessary—I had a nonterminal for SEMI but removed it (so I needed the annotation to give this the same precedence as SEMI).

They're mostly not necessary, but they resolve shift/reduce conflicts in some cases. The major one we have is one for `if`, to handle the famous "dangling else" problem. There's a section of the Menhir manual about conflicts that's really helpful!